Update azure_rm_storageaccount.py

paultaiton opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Fix passing minimum_tls_version, allow_blob_public_access, and https_only on new storage account creation.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #457 
Fixes #325 
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
azure_rm_storageaccount

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
This is a bugfix in the creation of new storage account, and update of tests to catch this bug in future regressions.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

paultaiton committed:
Update azure_rm_storageaccount.py

paultaiton commented:
@Fred-sun,
I believe you wrote the original code that I'm modifying in this PR. Your review would be appreciated.

@arsenicks,
This should correct the problem you have in the short term, until I can complete the more comprehensive PR.

arsenicks commented:
@paultaiton Ohhh thanks a lot for this!!

Fred-sun commented on plugins/modules/azure_rm_storageaccount.py:
```diff
@@ -823,7 +823,7 @@ def create_account(self):
                 account_dict['blob_cors'] = self.blob_cors
             return account_dict
-        if bool(self.https_only):
+        if not bool(self.https_only):
```
```suggestion
        if self.https_only is None:
```

Fred-sun commented on plugins/modules/azure_rm_storageaccount.py:
```diff
@@ -823,7 +823,7 @@ def create_account(self):
             return account_dict
-        if bool(self.https_only):
+        if not bool(self.https_only):
             self.https_only = False
```
```suggestion
            self.https_only = True
```

Fred-sun reviewed (commented):
@paultaiton I have checked the parameter in azure-cli, https-only default value is true. Please update follow the comment and update the doc---line 90. Thank you very much!

https://github.com/ansible-collections/azure/blob/bb56aec4b2c8ecd71ebb8ab0267f575967b8dedd/plugins/modules/azure_rm_storageaccount.py#L90
```e.g.
link: https://docs.microsoft.com/en-us/cli/azure/storage/account?view=azure-cli-latest#az_storage_account_create

paultaiton committed:
Set https only to True by default

Changed https_only parameter to be True by default, and removed the conditional to convert from None type.

paultaiton committed:
undo line deletion.

paultaiton commented:
If we are setting https_only parameter to be True by default (which makes sense to match storageaccount API updates 2019-06-01,) then the line 826 
```
         if not bool(self.https_only): 
             self.https_only = False 
``` 

is no longer necessary, since the property can only ever be True or False, and never None.

Fred-sun commented:
> If we are setting https_only parameter to be True by default (which makes sense to match storageaccount API updates 2019-06-01,) then the line 826
> 
> ```
>          if not bool(self.https_only): 
>              self.https_only = False 
> ```
> 
> is no longer necessary, since the property can only ever be True or False, and never None.
> 
> I've modified the code to match this convention instead.

Yes! agree with your comments!

Fred-sun commented on plugins/modules/azure_rm_storageaccount.py:
```diff
@@ -87,7 +87,7 @@
             - Allows https traffic only to storage service when set to C(true).
             - Allows update storage account property when set to C(False).
-            - Default value is C(False).
+            - Default value is C(True).
```
```suggestion
        default: True
```

Fred-sun reviewed (commented):
Please

paultaiton committed:
Update plugins/modules/azure_rm_storageaccount.py

Co-authored-by: Fred-sun <37327967+Fred-sun@users.noreply.github.com>

paultaiton committed:
tls and blob public

paultaiton committed:
Merge branch 'paultaiton_20210315_storage-https' of github.com:paultaiton/azure into paultaiton_20210315_storage-https

paultaiton commented:
@Fred-sun 
I added a couple more changes since I noticed that minimum_tls_version had the same bug, and to make blob_public_access more explicit and obvious with the defaults.

Fred-sun commented on plugins/modules/azure_rm_storageaccount.py:
```diff
@@ -87,23 +87,23 @@
             - If set to false, no containers in this account will be able to allow anonymous public access.
-            - Default value is C(True).
         type: bool
+        default: True
```
not need! The default value for this property is null

Fred-sun reviewed (commented):
Please!

Fred-sun commented:
@paultaiton Please update the follow lines,  If set minimum_tls_version default value!, It will update the minimum_tls_version value when update account blob public access.  advice change to '- output.storageaccounts[0].minimum_tls_version == "TLS1_0" '
https://github.com/ansible-collections/azure/blob/553cbd277fb84488fa11a1a057efdfc44b99412b/tests/integration/targets/azure_rm_storageaccount/tasks/main.yml#L190

arsenicks commented:
> @Fred-sun
> I added a couple more changes since I noticed that minimum_tls_version had the same bug, and to make blob_public_access more explicit and obvious with the defaults.

I was aware of it and can confirm it's not working as expected in 1.4 but didn't want to bother you with this in the short term as you were working on the whole thing :P Anyway thanks for fixing it, let me know if you need any kind of test. We have a policy in place that was blocking the creation because of the default ACL not being Deny, my first debug lead me to think this is not the same problem but if you can take a second to lookup that part of the code and see if it's similar that would be great.

I'll create another issue for the ACL when I have more information, it's not clear what cause the problem right now.

One question for you guy's, not putting any pressure, just want to know how released are managed with the collection. Do you ship a new collection version on a specific "schedule" or just as needed ? If so is there a planned released in the coming weeks ?

Thanks a lot to you both!

Fred-sun commented:
@arsenicks There will probably be an edition of Collections in a month or two.

paultaiton committed:
Merge remote-tracking branch 'upstream/dev' into paultaiton_20210315_storage-https

Fred-sun commented:
@haiyuazhang

paultaiton committed:
Merge remote-tracking branch 'upstream/dev' into paultaiton_20210315_storage-https

paultaiton committed:
remove defaults

paultaiton committed:
Merge branch 'paultaiton_20210315_storage-https' of github.com:paultaiton/azure into paultaiton_20210315_storage-https

paultaiton committed:
gitignore for secret information

Fred-sun commented on plugins/modules/azure_rm_storageaccount.py:
```diff
@@ -822,12 +821,6 @@ def create_account(self):
-            self.https_only = False
-        if bool(self.minimum_tls_version):
-            self.minimum_tls_version = 'TLS1_0'
-        if bool(self.allow_blob_public_access):
```
if self.allow_blob_public_access is None:
            self.allow_blob_public_access = True

Fred-sun commented on plugins/modules/azure_rm_storageaccount.py:
```diff
@@ -822,12 +821,6 @@ def create_account(self):
             return account_dict
-        if bool(self.https_only):
-            self.https_only = False
-        if bool(self.minimum_tls_version):
```
if self.minimum_tls_version is None:
            self.minimum_tls_version = 'TLS1_0'

Fred-sun reviewed (commented):
Please!

paultaiton commented:
Hey @Fred-sun ,
The test that failed is correct, it uncovered a bug I introduced.
Luckily I finally figured out how to get these tests running locally, so I'm updating both the integration tests and my code to account for omitting parameters while updating existing accounts. Please hold off on reviewing anything until I comment again, as I've corrected many of the bugs that haven't been pushed up yet.

paultaiton committed:
update tests

paultaiton committed:
change tests and docs

paultaiton committed:
undo

paultaiton commented:
@Fred-sun 

OK, I have reorganized and updated the tests to include more checks that should catch similar bugs in the future. It now creates and updates two storage accounts using implicit defaults and explicit settings for creating new accounts and updating existing accounts.

I added to documentation to explain what happens during parameter omission, but left it to use the API (also matches CLI) behavior.

I've run the integration tests and everything was successful, so I think it's ready for review for everything within the scope of this immediate fix. I did uncover a problem with the returned values for network_acls and blob_cors (integration test lines 227-229 that are commented out,) but it's outside the scope of the code I've touched, and I'm not familiar enough with its operation to fix it within this PR. I'll try and get to it later if I can.

Appreciate your help Fred.

l3ender commented:
@paultaiton Would you mind sharing detail on how you were able to get tests running locally? I've been trying to do the same but haven't had luck. Any detail you can share on running specific tests and the test suite would be appreciated!

paultaiton commented:
> @paultaiton Would you mind sharing detail on how you were able to get tests running locally? I've been trying to do the same but haven't had luck. Any detail you can share on running specific tests and the test suite would be appreciated!

Yes, absolutely. I was already planning on making a Documentation PR that describes exactly how to set up an environment for linting and integration tests.

Fred-sun commented on plugins/modules/azure_rm_storageaccount.py:
```diff
@@ -706,8 +705,7 @@ def update_account(self):
-        if self.allow_blob_public_access is not None:
-            if bool(self.allow_blob_public_access) != bool(self.account_dict.get('allow_blob_public_access')):
+        if self.allow_blob_public_access is not None and self.allow_blob_public_access != self.account_dict.get('allow_blob_public_access'):
                 self.results['changed'] = True
```
@paultaiton  over indented! Thanks!

paultaiton replied:
Got it @Fred-sun . Somehow my linter got disabled. Thanks.

paultaiton committed:
whitespace alignment.

paultaiton commented:
Thanks again for your help @haiyuazhang and @Fred-sun . Your support is appreciated.

@arsenicks please let me know if you see any trouble after switching to this new module update.

l3ender commented:
@paultaiton Is there any detail you can share in the meantime on how to set up an environment for linting and integration tests? Being able to lint/test on my local environment would help tremendously. Once getting it set up, I'm happy to create and/or contribute to a documentation PR.

paultaiton commented:
I use pycodestyle in vscode with --max-line-length=160       and
--ignore=E###,W###

As command line arguments (settings in python extension.) I forget the E number, but the one for module imports not at top of file, and the warn for operators at start of new line. The latter is acceptable according to PEP8, so I consider it a fault of the pycodestyle lint.

The ansible-test bit I'll have to get you when I'm back home, as there is a cloud config file you need to define that I can't recall from memory.
